### ISSUE TITLE
Feature request: Unpublish when deleting resources

### ISSUE THREAD

**Issue by mrcycling** [modxcms/revolution#12924]
Since deleting resources means removing them from the website, it makes sense that they should also be unpublished at the same time as they are deleted.  This would save having to apply extra filters to removed deleted but still published resources in plugins, extras and other such functions.

Some thoughts I had in regard to maintaining the reversibility currently enjoyed with deletes:

OnDelete
	check if resource is published
	if published, unpublish and set unpub_date identical to deletedon date
	if not published, change nothing

`// wild guess at code changes`
`if ($this->resource->get('published') == 1) {`
`$this->resource->set('published',false);`
`$this->resource->set('unpub_date',$this->deletedTime);`
`} else { }`

OnUnDelete
	compare upub_date and deletedon date
	if the same, publish resource
	if not the same, change nothing

`// 2nd wild guess at code`
`if ($this->resource->get('deletedon') == $this->resource->get('unpub_date')) {`
`$this->resource->set('published',true);`
`$this->resource->set('unpub_date',0);`
`} else { }`

In both cases do not change pub_date or editedon, leave them as is to reflect the most recent official dates prior to the 'oops i deleted the wrong damn resource'.

**Comment by OptimusCrime** [comment]
This can easily be done with plugins if you wish.

Also, I disagree with this behavior. If I delete and then undelete a resource I expect it to work like it did before I deleted it. If this behavior is added I would also have to publish the resource afterwards. In my opinion are deleted resources simply deleted. Their published status is irrelevant.

**Comment by mrcycling** [comment]
As per my suggestion, the published status would be restored upon
undeleting the resource. There would be no need to manually republish.

**Comment by pixelchutes** [comment]
What if the Resource was unpublished prior to being deleted?

:thumbs_down: I agree with @OptimusCrime, we should be careful making assumptions on behalf of the user.

**Comment by mrcycling** [comment]
My apologies if my logic was not understandable, let me try again .... I am assuming the magic takes place in core/model/modx/processors/resource and in the delete.class.php and undelete.class.php.

In delete.class.php around line 70ish, one could test for published status of the resource being deleted.  If the resource is published, then set published status to 0/false and set the upublished date (unpub_date) to the same time stamp that gets assigned to the deleted on date (deletedon). If the resource is unpublished then do not change the published and unpub_date values.

In undelete.class.php around line 40ish, you would test to see if the unpublished date (unpub_date) is the same as the deleted on date (deletedon).  If the two dates are identical, then you know that resource was published prior to deletion, so then set the published state to 1/true and the unpub_date value to 0/null, same as you do for the deletedon value.  If the dates do not match, then you know the resource was unpublished prior to delting so leave the published / unpub_date vaues alone as they already reflect the original state.This way the restored deleted resource has the exact same setting for publishing as it had before it was temporarily deleted.

Can you guys provide any practical examples of when a developer would want a resource that is published AND deleted?  (other than the aforementioned undelete function)  On the other hand there are a myriad of functions using published resources that don't want one that is deleted.

It seems logical that the default conditions should be those that work for the majority of uses and that the oddball usage should bear the burden of additional filtering/coding.  Currently it seems the majority of uses have to do extra filtering on the grounds that maybe someday someone will need resources that are deleted AND published.

**Comment by mrcycling** [comment]
I guess I don't think inside the MODX box,so am closing my request.

**Comment by sottwell** [comment]
No, you are missing the point. MODX is not a box. This sort of thing is exactly what plugins are for.

**Comment by mrcycling** [comment]
@Susan my reference to thinking inside (or outside) the modx box was about
the mindset of modx, not about limitations.  Obviously my ideas of the way
things should work are quite different than modx. Case in point, deleting
... I feel it is removal, you folks look at it more like sleeping. So
really it is exactly what you mentioned in the forum, being told how things
should work.

**Comment by jpdevries** [comment]
There are several plugins like VersionX that rely on "temporary" hidden versions of resources. Such an assumption could screw the ways those plugins work up because they may delete and undelete "versions" routinely. 
The assumption makes perfect sense, for some use cases, but even still probably doesn't belong in the core. 
A plugin is your best bet.

**Comment by sebastian-marinescu** [comment]
I also see MODX more as a CMF with which you develop your tailor-made CMS.

**Comment by mrcycling** [comment]
@jpdevries I appreciate your response citing examples, rather than the I :heart: MODX mantra. And I think you inadvertently nailed the issue. 

Dictionaries defines "delete" as to remove, to erase, to eliminate.  And I would be willing to bet that 99% of English speaking computer users (and probably non-English speakers as well) have that same understanding of that function.  It is one of the most universal computer concepts ... save ... delete.

What MODX has is more like a HIDE, the resource is hidden from outside browsers but still active inside MODX.  It is not a "delete" in any sense of the word.

Perhaps MODX needs to rename the current delete to **Hide** and then create a proper **Delete** function, so that everyone from the developer right down to Betty Boop Business Woman & end user automatically understands what it means.

Yes it means some developers will need to revisit their code and perhaps change OnDeleteXX events to OnHideXX events.  And yes it may mean that in some cases someone might have to Hide and then Delete and then Empty Trash or perhaps add a modal on the trash can to offer options to remove hidden, deleted or all.  But ultimately the end user will understand better and be happier, hence the developer will be happier, hence the creators of MODX will be happier, cause smiles roll uphill.

**Comment by sottwell** [comment]
Strikethrough has always meant "flag for delete" in word processing either by hand, typewriter or electronically, and the concept of "delete" meaning "put in the trash" for files has also been around for as long as there has been a GUI.

If you don't like the specific term "delete" you can very easily change the lexicon. That is why we have events and plugins and lexicon files. 

That way MODX doesn't have to try to please everybody, everybody can please themselves without forcing their particular views on anybody else - which is the definition of "freedom", after all. And even more smiles all around, no?

The whole point here is, no matter what you want to call it, "delete", "trash", "hide", "blue-done-fuzzy", it is not "unpublish", it is a completely different status.
